### PR TITLE
13: add User Profile page

### PR DESCRIPTION
**Pull request by serhiiyasenev**
This PR instead https://github.com/BinaryStudioAcademy/bsa-2023-squirrel/pull/64

image

### CODE REVIEW

**Review by VladTkach — COMMENTED** [review]
I don't think we need UserDto in every components where we use user info, because some property such a emailNotification or isGoogleAuth used only in profile page, you can add just some enpoint to get all info about user. so it's beter have. Better to have more user models than dragging unnecessary information everywhere

**Comment by serhiiyasenev** [comment]
UserProfileDto was added

File: backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs [hunk 1]
```diff
@@ -54,7 +71,82 @@ public async Task<User> CreateUserAsync(UserRegisterDto userDto, bool isGoogleAu
...
+        return _mapper.Map<UserDto>(createdUser); ;
+    }
+
+    public async Task<UserProfileDto> UpdateUserNamesAsync(UpdateUserNamesDto updateUserDTO)
+    {
+        var userEntity = await GetUserByIdInternal(updateUserDTO.Id);
+
+        if (await GetUserByUsernameInternal(updateUserDTO.Username) is not null)
+        {
+            throw new UsernameAlreadyRegisteredException();
```
Comment (VladTkach) [inline backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs:92]: what this check is used for?
I just try to change FirstName and received error
image

File: frontend/src/app/modules/user-profile/user-profile-page/user-profile.component.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                    Validators.maxLength(25),
+                    ValidationsFn.lowerExist(),
+                    ValidationsFn.upperExist(),
+                ],
+            ],
+            newPassword: [
+                '',
+                [
+                    Validators.required,
+                    Validators.minLength(2),
```
Comment (VladTkach) [inline frontend/src/app/modules/user-profile/user-profile-page/user-profile.component.ts]: Acording to specification we need at lest 6 character in the password, just so as not to rework later

File: backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs [hunk 3]
```diff
@@ -54,7 +71,82 @@ public async Task<User> CreateUserAsync(UserRegisterDto userDto, bool isGoogleAu
...
+    {
+        var userEntity = await _context.Users.FirstOrDefaultAsync(u => u.Email == email);
+        return userEntity;
+    }
+
+    public async Task<User?> GetUserEntityByUsername(string username)
+    {
+        var userEntity = await _context.Users.FirstOrDefaultAsync(u => u.Username == username);
+        return userEntity;
+    }
```
Comment (OlegKarapysh) [inline backend/Squirrel.Core/Squirrel.Core.BLL/Services/UserService.cs]: no need for a variable `userEntity`

File: backend/Squirrel.Shared/Exceptions/EntityNotFoundException.cs [hunk 4]
```diff
@@ -1,8 +1,36 @@
...
+        $"Entity '{entityName}' with id '{id}' not found",
+        ErrorType.NotFound,
+        HttpStatusCode.NotFound)
+    {
+    }
+
+    public EntityNotFoundException(string entityName, string property) : base(
+        $"Entity '{entityName}' with id '{property}' not found",
+        ErrorType.NotFound,
+        HttpStatusCode.NotFound)
```
Comment (OlegKarapysh) [inline backend/Squirrel.Shared/Exceptions/EntityNotFoundException.cs]: seems like incorrect message

File: frontend/src/app/core/services/user.service.ts [hunk 5]
```diff
@@ -1,20 +1,38 @@
...
+import { UserProfileDto } from 'src/app/models/user/user-profile-dto';
 
 import { HttpInternalService } from './http-internal.service';
 
-@Injectable({
-    providedIn: 'root',
-})
+@Injectable({ providedIn: 'root' })
 export class UserService {
     private routePrefix = '/api/user';
```
Comment (OlegKarapysh) [inline frontend/src/app/core/services/user.service.ts]: make `routePrefix` readonly, it doesn't need to be changed

File: frontend/src/app/modules/user-profile/user-profile-routing.module.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { NgModule } from '@angular/core';
+import { RouterModule, Routes } from '@angular/router';
+
+import { UserProfileComponent } from './user-profile-page/user-profile.component';
+
+const routes: Routes = [
+    {
+        path: '',
+        component: UserProfileComponent,
+        children: [],
```
Comment (OlegKarapysh) [inline frontend/src/app/modules/user-profile/user-profile-routing.module.ts]: seems like we don't need to specify `children:` here

**Review by VladTkach — APPROVED** [review]

**Review by Anatolii-Ivanov — APPROVED** [review]

**Pull request merged**